sci-libs/pytorch/pytorch-1.4.0.ebuild

```diff
@@ -0,0 +1,253 @@
+	"${FILESDIR}/0002-Don-t-build-libtorch-again-for-PyTorch-1.4.0.patch"
+	"${FILESDIR}/0003-Change-path-to-caffe2-build-dir-made-by-libtorch.patch"
+	"${FILESDIR}/0004-Don-t-fill-rpath-of-Caffe2-library-for-system-wide-i.patch"
+	"${FILESDIR}/0005-Change-library-directory-according-to-CMake-build.patch"
+	"${FILESDIR}/0006-Change-torch_path-part-for-cpp-extensions.patch"
+	"${FILESDIR}/0007-Add-necessary-include-directory-for-ATen-CPU-tests.patch"
+	"${FILESDIR}/0008-Fix-include-directory-variable-of-rocThrust-1.4.0.patch"
+	"${FILESDIR}/0009-Prefer-lib64-for-some-ROCm-packages-1.4.0.patch"
+	"${FILESDIR}/0010-Remove-conversion-ambiguity-in-ternary-operators.patch"
+	"${FILESDIR}/0011-Prevent-finding-pybind11-system-wide-installation.patch"
+	"${FILESDIR}/0012-Special-path-for-roctracer.patch"
+)
+
+src_prepare() {
+	cmake-utils_src_prepare
+
+	if use rocm; then
+		${EPYTHON} "${S}/tools/amd_build/build_amd.py"
+	fi
+}
+
+src_configure() {
+	local blas="Eigen"
+
+	if use atlas; then
+		blas="ATLAS"
+	elif use mkl; then
+		blas="MKL"
+	elif use openblas; then
+		blas="OpenBLAS"
+	fi
+
+	if use rocm; then
+		export HCC_PATH="${HCC_HOME}"
+		export ROCBLAS_PATH="/usr"
+		export ROCFFT_PATH="/usr"
+		export HIPSPARSE_PATH="/usr"
+		export HIPRAND_PATH="/usr"
+		export ROCRAND_PATH="/usr"
+		export MIOPEN_PATH="/usr"
+		export RCCL_PATH="/usr"
+		export ROCPRIM_PATH="/usr"
+		export HIPCUB_PATH="/usr"
+		export ROCTHRUST_PATH="/usr"
+		export ROCTRACER_PATH="/usr"
+	fi
+
+	local mycmakeargs=(
+		-DTORCH_BUILD_VERSION=${PV}
+		-DTORCH_INSTALL_LIB_DIR=lib64
+		-DBUILD_BINARY=$(usex tools ON OFF)
+		-DBUILD_CUSTOM_PROTOBUF=OFF
+		-DBUILD_DOCS=$(usex doc ON OFF)
+		-DBUILD_PYTHON=$(usex python ON OFF)
+		-DBUILD_SHARED_LIBS=$(usex static OFF ON)
+		-DBUILD_TEST=$(usex test ON OFF)
+		-DUSE_ASAN=$(usex asan ON OFF)
+		-DUSE_CUDA=$(usex cuda ON OFF)
+		-DUSE_ROCM=$(usex rocm ON OFF)
+		-DUSE_FBGEMM=$(usex fbgemm ON OFF)
+		-DUSE_FFMPEG=$(usex ffmpeg ON OFF)
+		-DUSE_GFLAGS=$(usex gflags ON OFF)
+		-DUSE_GLOG=$(usex glog ON OFF)
+		-DUSE_LEVELDB=$(usex leveldb ON OFF)
+		-DUSE_LITE_PROTO=OFF
+		-DUSE_LMDB=$(usex lmdb ON OFF)
+		-DCAFFE2_USE_MKL=$(usex mkl ON OFF)
+		-DUSE_MKLDNN=$(usex mkldnn ON OFF)
+		-DUSE_MKLDNN_CBLAS=OFF
+		-DUSE_NCCL=OFF
+		-DUSE_NNPACK=$(usex nnpack ON OFF)
+		-DUSE_NUMPY=$(usex numpy ON OFF)
+		-DUSE_NUMA=$(usex numa ON OFF)
+		-DUSE_OBSERVERS=$(usex observers ON OFF)
+		-DUSE_OPENCL=$(usex opencl ON OFF)
+		-DUSE_OPENCV=$(usex opencv ON OFF)
+		-DUSE_OPENMP=$(usex openmp ON OFF)
+		-DUSE_TBB=OFF
+		-DUSE_PROF=OFF
+		-DUSE_QNNPACK=$(usex qnnpack ON OFF)
+		-DUSE_REDIS=$(usex redis ON OFF)
+		-DUSE_ROCKSDB=OFF
+		-DUSE_ZMQ=$(usex zeromq ON OFF)
+		-DUSE_MPI=$(usex mpi ON OFF)
+		-DUSE_GLOO=$(usex gloo ON OFF)
+		-DBUILD_NAMEDTENSOR=$(usex namedtensor ON OFF)
+		-DBLAS=${blas}
+		-DBUILDING_SYSTEM_WIDE=ON # to remove insecure DT_RUNPATH header
+	)
+
+	local CC="clang"
+	local CXX="clang++"
```

Comment:
Why is clang required?  Is it because of ROCm?

Reply:
No, not really, it's because of the code in version 1.4.0, in `c10`, if I remember correctly (haven't saved the failure log for some reason), which interferes with the [regression in GCC 9.1 and 9.2](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90538). It's fixed in 9.3 and 10 and doesn't reproduce in 8.x and Clang, hence the change. It can be more fine-grained, allowing safe GCC versions, but require detecting compiler version in ebuild, which I'm not so comfortable with.

Reply:
I will revert this.  We cannot switch to clang unconditionally.

Reply:
Then it's probably better not to include ebuild for version 1.4.0 for now, as otherwise it would lead to compilation failure, as soon as GCC-9 is selected as default compiler (which is the default for modern profiles).

Reply:
I can't reproduce this with `gcc-9.2.0-r4` though.

Reply:
It's fairly possible, especially given, that, as I can see, the merged version of the ebuild have little similar parts to one in this request. I think, I have `gcc-9.2.0-r2`, the latest stable version, but haven't tested building the merged version, yet.